storage/filesystem/object.go

```diff
@@ -71,7 +71,14 @@ func (s *ObjectStorage) Get(h core.Hash) (core.Object, error) {
 	p := packfile.NewParser(r)
 
 	obj := s.NewObject()
-	return obj, p.FillObject(obj)
+	err = p.FillObject(obj)
+	if err != nil {
+		return nil, err
+	}
+	if core.AnyObject != t && obj.Type() != t {
```

Comment:
Calling p.ReadObjectTypeAndLength before would complicate FillObject logic a little bit though, because it would consume from the input reader, so either FillObject would need to take object type and length as parameters, or ReadObjectTypeAndLength should cache the result for multiple invocations.